File: backend/app/schemas/simulation.py
from __future__ import annotations

from datetime import datetime
from typing import Any, List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel


class SimulationStartRequest(BaseModel):
    scenario: str = "normal"


class SimulationStartResponse(BaseModel):
    status: str
    scenario: str
    speed: int


class SimulationPauseResponse(BaseModel):
    status: str


class SimulationSpeedRequest(BaseModel):
    multiplier: Literal[1, 5, 10]


class SimulationSpeedResponse(BaseModel):
    speed: int


class SimulationScenarioRequest(BaseModel):
    scenario: str


class SimulationScenarioResponse(BaseModel):
    scenario: str
    status: str


class SimulationTriggerAnomalyRequest(BaseModel):
    campaign_id: UUID


class SimulationTriggerAnomalyResponse(BaseModel):
    triggered: bool
    campaign_id: UUID


class SimulationResetResponse(BaseModel):
    status: str
    message: str


class SimulationStatusResponse(BaseModel):
    status: str
    sim_time: Optional[datetime] = None
    speed: int
    scenario: str
    tick_count: int
    campaign_count: int


class SimulationLogEntry(BaseModel):
    id: UUID
    event_type: str
    campaign_id: Optional[UUID] = None
    sim_time: datetime
    real_time: datetime
    details: Optional[dict[str, Any]] = None


class SimulationLogResponse(BaseModel):
    entries: List[SimulationLogEntry]
    total: int
    limit: int
    offset: int

File: frontend/src/api/simulation.ts
import apiClient from './client'
import type {
  SimulationStartResponse,
  SimulationPauseResponse,
  SimulationSpeedResponse,
  SimulationScenarioResponse,
  SimulationTriggerAnomalyResponse,
  SimulationResetResponse,
  SimulationStatusResponse,
  SimulationLogResponse,
} from './types'

export async function startSimulation(scenario?: string): Promise<SimulationStartResponse> {
  const { data } = await apiClient.post<SimulationStartResponse>('/simulation/start', {
    scenario: scenario ?? 'normal',
  })
  return data
}

export async function pauseSimulation(): Promise<SimulationPauseResponse> {
  const { data } = await apiClient.post<SimulationPauseResponse>('/simulation/pause')
  return data
}

export async function setSpeed(multiplier: 1 | 5 | 10): Promise<SimulationSpeedResponse> {
  const { data } = await apiClient.put<SimulationSpeedResponse>('/simulation/speed', { multiplier })
  return data
}

export async function setScenario(scenario: string): Promise<SimulationScenarioResponse> {
  const { data } = await apiClient.put<SimulationScenarioResponse>('/simulation/scenario', { scenario })
  return data
}

export async function triggerAnomaly(campaignId: string): Promise<SimulationTriggerAnomalyResponse> {
  const { data } = await apiClient.post<SimulationTriggerAnomalyResponse>('/simulation/trigger-anomaly', {
    campaign_id: campaignId,
  })
  return data
}

export async function resetSimulation(): Promise<SimulationResetResponse> {
  const { data } = await apiClient.post<SimulationResetResponse>('/simulation/reset')
  return data
}

export async function fetchStatus(): Promise<SimulationStatusResponse> {
  const { data } = await apiClient.get<SimulationStatusResponse>('/simulation/status')
  return data
}

export async function fetchLog(params?: {
  limit?: number
  offset?: number
  event_type?: string
}): Promise<SimulationLogResponse> {
  const { data } = await apiClient.get<SimulationLogResponse>('/simulation/log', { params })
  return data
}
